### PR TITLE
Issue 6476 bob exception reports

### LINKED ISSUE #6476: Bob should report better message on engine build error
Sometimes the logging code in bob.jar throws an error, and it's hard to pinpoint the issue.

https://forum.defold.com/t/i-get-the-resource-cannot-be-null-error-in-the-latest-version-of-the-engine/70503/35?u=mathias_westerdahl

1. When the "resource" is null, we should log the error in full
2. We should fix an ecplicit error output/validation for the package names

### PR DESCRIPTION
**Pull request by JCash**

### CODE REVIEW

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Bob.java [hunk 1]
```diff
@@ -706,7 +706,13 @@ private static void mainInternal(String[] args) throws IOException, CompileExcep
...
                 errors.append(logExceptionToString(info.getSeverity(), info.getResource(), info.getLineNumber(), info.getMessage()) + "\n");
             }
-            errors.append("\nFull log: \n" + e.getRawLog() + "\n");
+
+            String msg = String.format("For the full log, see %s (or add -v)\n", e.getLogPath());
+            errors.append(msg);
+
+            if (verbose) {
+                errors.append("\nFull log: \n" + e.getRawLog() + "\n");
+            }
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Bob.java:715]: Make it an option (previously we always output all the error + the log)

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Bob.java [hunk 2]
```diff
@@ -706,7 +706,13 @@ private static void mainInternal(String[] args) throws IOException, CompileExcep
             {
                 errors.append(logExceptionToString(info.getSeverity(), info.getResource(), info.getLineNumber(), info.getMessage()) + "\n");
             }
-            errors.append("\nFull log: \n" + e.getRawLog() + "\n");
+
+            String msg = String.format("For the full log, see %s (or add -v)\n", e.getLogPath());
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Bob.java:710]: Reference the log file, so it becomes better known.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Bob.java [hunk 3]
```diff
@@ -745,15 +751,26 @@ private static void mainInternal(String[] args) throws IOException, CompileExcep
...
-            System.err.println(e.getMessage());
-            System.err.println("Cause: " + e.getCause());
-            e.printStackTrace();
-            System.exit(1);
+        } catch (LibraryException|CompileExceptionError e) {
+            logErrorAndExit(e);
+        } catch (Exception e) {
+            logErrorAndExit(e);
         }
+
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Bob.java:773]: Also, we don't want it to look too messy if we get random errors.
We enable the traceback on "-v"

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/CompileExceptionError.java [hunk 4]
```diff
@@ -52,6 +52,12 @@ public CompileExceptionError(Throwable e) {
         this.lineNumber = -1;
     }
 
+    public CompileExceptionError(String message) {
+        super(message);
+        this.resource = null;
+        this.lineNumber = -1;
+    }
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/CompileExceptionError.java:59]: Allow the exception to receive just a string.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/MultipleCompileException.java [hunk 5]
```diff
@@ -98,6 +99,14 @@ public void attachLog(IResource contextResource, String rawLog) {
         this.rawLog = rawLog;
     }
 
+    public void setLogPath(String path) {
+        this.logPath = path;
+    }
+
+    public String getLogPath() {
+        return this.logPath;
+    }
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/MultipleCompileException.java:108]: We'll use this path when reporting in bob/editor

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Project.java [hunk 6]
```diff
@@ -616,6 +616,9 @@ public void mount(IResourceScanner resourceScanner) throws IOException, CompileE
         this.fileSystem.clearMountPoints();
         this.fileSystem.addMountPoint(new ClassLoaderMountPoint(this.fileSystem, "builtins/**", resourceScanner));
         Map<String, File> libFiles = LibraryUtil.collectLibraryFiles(getLibPath(), this.libUrls);
+        if (libFiles == null) {
+            throw new CompileExceptionError("Missing libraries folder. You need to run the 'resolve' command first!");
+        }
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Project.java:621]: Another issue: If you try to build before resolving the libraries, you got an ugly nulll pointer callstack, with no info about the problem.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/BundleHelper.java [hunk 7]
```diff
@@ -1131,7 +1131,12 @@ public static File buildEngineRemote(ExtenderClient extender, String platform, S
                     buildError = FileUtils.readFileToString(logFile);
                     parseLog(platform, buildError, issues);
                     MultipleCompileException exception = new MultipleCompileException("Build error", e);
+                    IResource projectResource = project.getGameProjectResource();
                     IResource extManifestResource = ExtenderUtil.getResource(allSource.get(0).getPath(), allSource);
+                    if (extManifestResource == null) {
+                        extManifestResource = projectResource;
+                    }
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/BundleHelper.java:1138]: The actual issue: If the fallback resource was null, then we got an exception further down, messing up the error reporting.

File: editor/src/clj/editor/engine/build_errors.clj [hunk 8]
```diff
@@ -484,15 +484,25 @@
 (defn- compile-exception-error-causes [project evaluation-context exception]
   [(error-info-provider->cause project evaluation-context exception)])
 
+(defn- output-log-path [project evaluation-context ^MultipleCompileException exception]
+  (let [log-path (.getLogPath exception)]
+    [(g/map->error
+      {:_node-id nil ;; The editor cannot currently reference files in the /build folder
+       :message (str "For the full log, see " log-path)
+       :severity :warning})]))
+
```
Comment (JCash) [inline editor/src/clj/editor/engine/build_errors.clj:493]: Editor cannot currently reference files in the build folder :(

File: editor/src/clj/editor/engine/build_errors.clj [hunk 9]
```diff
@@ -484,15 +484,25 @@
...
                             evaluation-context)]
-    (or (try-parse-invalid-lib-error-causes project evaluation-context log)
-        (try-parse-compiler-error-causes project evaluation-context log ext-manifest-file)
-        (generic-extension-error-causes project evaluation-context log))))
+    (into [] (concat
+              (or (try-parse-invalid-lib-error-causes project evaluation-context log)
+                  (try-parse-compiler-error-causes project evaluation-context log ext-manifest-file)
+                  (generic-extension-error-causes project evaluation-context log))
+              
+              (output-log-path project evaluation-context exception)))))
```
Comment (JCash) [inline editor/src/clj/editor/engine/build_errors.clj:505]: Output the errors and the extra info about the log file

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Bob.java [hunk 10]
```diff
@@ -745,15 +751,26 @@ private static void mainInternal(String[] args) throws IOException, CompileExcep
...
         try {
             mainInternal(args);
-        } catch (LibraryException e) {
-            System.err.println(e.getMessage());
-            System.err.println("Cause: " + e.getCause());
-            e.printStackTrace();
-            System.exit(1);
+        } catch (LibraryException|CompileExceptionError e) {
+            logErrorAndExit(e);
+        } catch (Exception e) {
```
Comment (britzl) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Bob.java:770]: Might as well only catch Exception and not any of the others? I guess Exception is the base class for our own exceptions?

**Review by britzl — DISMISSED** [review]
Looks good

**Review by britzl — APPROVED** [review]

**Pull request merged**